mlflow/tensorflow.py

```diff
@@ -619,12 +619,19 @@ def _flush_queue():
         # flush operation should proceed; all others are redundant and should be dropped
         acquired_lock = _metric_queue_lock.acquire(blocking=False)
         if acquired_lock:
-            global _metric_queue
             client = mlflow.tracking.MlflowClient()
-            dic = _assoc_list_to_map(_metric_queue)
-            for key in dic:
-                try_mlflow_log(client.log_batch, key, metrics=dic[key], params=[], tags=[])
-            _metric_queue = []
+            # For thread safety and to avoid modifying a list while iterating over it, we record a
+            # separate list of the items being flushed and remove each one from the metric queue,
+            # rather than clearing the metric queue or reassigning it (clearing / reassigning is
+            # dangerous because we don't block threads from adding to the queue while a flush is
+            # in progress)
+            snapshot = _metric_queue[:]
+            for item in snapshot:
+                _metric_queue.remove(item)
+
+            metrics_by_run = _assoc_list_to_map(snapshot)
+            for run_id, metrics in metrics_by_run.items():
+                try_mlflow_log(client.log_batch, run_id, metrics=metrics, params=[], tags=[])
```

Comment:
I am pretty sure this was the behavior before anyway, but this makes it a little more clear that if `try_mlflow_log` fails to log a metric x, that metric won't be in the queue anymore due to line 630 and so will never be logged.
Not blocking though since this was already the case.

Reply:
Yep! That's expected / is consistent with past behavior. Thanks for calling this out!